source/LiteralModel.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <mariana-trench/EventLogger.h>
+#include <mariana-trench/LiteralModel.h>
+#include <mariana-trench/Positions.h>
+
+namespace marianatrench {
+
+namespace {
+
+class LiteralModelConsistencyError {
+ public:
+  static void raise(const std::string& what) {
+    ERROR(1, "Literal Model Consistency Error: {}", what);
+    EventLogger::log_event("regex_model_consistency_error", what);
+  }
+};
+
+}
+
+static void check_taint_config_consistency(const std::string& pattern, const TaintConfig& frame) {
+  if (frame.kind() == nullptr) {
+    LiteralModelConsistencyError::raise(fmt::format(
+        "Model for regex `{}` must have a source kind.", pattern));
+  }
+}
+
+static Taint create_taint(const std::string& pattern, const std::vector<TaintConfig>& sources) {
+  Taint all_sources;
+  for (const auto& source : sources) {
+    mt_assert(source.is_leaf());
+    check_taint_config_consistency(pattern, source);
+    all_sources.join_with(Taint{std::move(source)});
+  }
+  return all_sources;
+}
+
+LiteralModel::LiteralModel() : regex_(pattern_) { }
+
+LiteralModel::LiteralModel(
+    std::string&& pattern,
+    const std::vector<TaintConfig>& sources)
+    : pattern_(std::move(pattern)), regex_(pattern_), sources_(create_taint(pattern_, sources)) {
+}
+
+LiteralModel::LiteralModel(const LiteralModel& other)
+    : pattern_(other.pattern_), regex_(pattern_), sources_(other.sources_) {
+}
+
+const std::string& LiteralModel::pattern() const {
+  return pattern_;
+}
+
+const Taint& LiteralModel::sources() const {
+  return sources_;
+}
+
+bool LiteralModel::matches(const std::string_view literal) const {
+  return re2::RE2::FullMatch(literal, regex_);
+}
+
+void LiteralModel::join_with(const LiteralModel& other) {
+  if (this != &other) {
+    sources_.join_with(other.sources_);
+  }
```

Comment:
If the patterns are different I think it would make sense to set it to null/empty?

Reply:
Now checking if they have the same pattern, resetting `regex_` otherwise.